Guard TensorPipe agent by USE_TENSORPIPE

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42682 Guard TensorPipe agent by USE_TENSORPIPE**

The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

lw committed:
Guard TensorPipe agent by USE_TENSORPIPE



Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

lw commented on torch/CMakeLists.txt:
```diff
@@ -264,6 +268,8 @@ target_link_libraries(torch_python ${TORCH_PYTHON_LINK_LIBRARIES})
 
 target_compile_definitions(torch_python PRIVATE ${TORCH_PYTHON_COMPILE_DEFINITIONS})
 
+target_compile_definitions(torch_python PUBLIC ${TORCH_PYTHON_PUBLIC_COMPILE_DEFINITIONS})
```
Not so happy about this, TBH.

The issue is that we guard some parts of the headers with USE_TENSORPIPE, which means that every downstream user that includes those headers must also set the right value of USE_TENSORPIPE. (Note that some "downstream users" are part of PyTorch itself, like the tests). Instead of asking each user to take care to set their USE_TENSORPIPE compile definition to the same value as their USE_TENSORPIPE CMake variable, we could do that for them by adding it to the interface of the CMake target.

We used to do this in TensorPipe itself, in order to allow users to detect what transports/channels had been enabled during compilation, but it came back to bite us, and it's the reason we didn't ship SHM and CMA in PyTorch 1.6. What we settled with now is to define these flags in an auto-generated header file.

However, I'm not aware of any such header file in PyTorch. And, moreover, this approach is what c10d itself is doing to let its users know which of Gloo/NCCL/MPI is enabled:
https://github.com/pytorch/pytorch/blob/fb8aa0046c554630cb9f173df8b4c7c751415312/torch/lib/c10d/CMakeLists.txt#L108-L118
So perhaps it's OK?

malfet replied:
You can ask cmake/buck to bake the dependency for you in one of generated header, for example by adding `#cmakedefine USE_TENSORPIPE` to https://github.com/pytorch/pytorch/blob/686705c98b43f65e48e3a920b24634e65f54aca1/c10/macros/cmake_macros.h.in#L7-L10

(Although, this header is not the perfect place for that,  but the same is true for other `.h.in` files)

lw replied:
Yes, I do agree that having a generated header is more robust, and I guess I'll do that. However, even though c10, ATen and Caffe2 have one, PyTorch doesn't have any generated header yet, so adding one might be a bit of work. I'd like to land this soon-ish as there's some other work blocked by it, so if it's acceptable I'd prefer to go with this for now and introduce the generated header later...

beauby commented on caffe2/CMakeLists.txt:
```diff
@@ -282,6 +282,8 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     add_library(tensorpipe_agent "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.cpp" "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.h")
     target_link_libraries(tensorpipe_agent PRIVATE torch c10d tensorpipe fmt::fmt-header-only)
     add_dependencies(tensorpipe_agent torch c10d tensorpipe)
+    target_compile_definitions(tensorpipe_agent PRIVATE USE_TENSORPIPE)
+    target_compile_definitions(tensorpipe_agent PUBLIC USE_TENSORPIPE)
```
Isn't the `PRIVATE` redundant with the following PUBLIC?

lw replied:
Good point, fixed it.

beauby commented on torch/CMakeLists.txt:
```diff
@@ -167,6 +169,8 @@ if(USE_DISTRIBUTED)
       list(APPEND TORCH_PYTHON_COMPILE_DEFINITIONS USE_C10D)
       if(USE_TENSORPIPE)
         list(APPEND TORCH_PYTHON_LINK_LIBRARIES tensorpipe)
+        list(APPEND TORCH_PYTHON_COMPILE_DEFINITIONS USE_TENSORPIPE)
+        list(APPEND TORCH_PYTHON_PUBLIC_COMPILE_DEFINITIONS USE_TENSORPIPE)
       endif()
```
Might be redundant as well.

lw replied:
Same.

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -282,6 +282,7 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     add_library(tensorpipe_agent "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.cpp" "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.h")
     target_link_libraries(tensorpipe_agent PRIVATE torch c10d tensorpipe fmt::fmt-header-only)
     add_dependencies(tensorpipe_agent torch c10d tensorpipe)
+    target_compile_definitions(tensorpipe_agent PUBLIC USE_TENSORPIPE)
```
Shouldn't this be guarded by `USE_TENSORPIPE`?
```suggestion
    if(USE_TENSORPIPE)
      add_library(tensorpipe_agent "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.cpp" "${TORCH_SRC_DIR}/csrc/distributed/rpc/tensorpipe_agent.h")
      target_link_libraries(tensorpipe_agent PRIVATE torch c10d tensorpipe fmt::fmt-header-only)
      add_dependencies(tensorpipe_agent torch c10d tensorpipe)
      target_compile_definitions(tensorpipe_agent PUBLIC USE_TENSORPIPE)
    endif()
```

lw replied:
Good catch, done! (In a slightly different way, as it's more convenient to have the `tensorpipe_agent` target define even when we don't use TensorPipe, even if it's empty).

malfet commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -1,5 +1,7 @@
 #include <torch/csrc/distributed/rpc/tensorpipe_agent.h>
 
+#ifdef USE_TENSORPIPE
+
```
Rather than having #ifdef guard, wouldn't it be better to exclude the file from the list of sources in torch/CMakeLists.txt?

lw replied:
I thought it would simpler (and more robust?) to add this guard in one place than to remove a file from the sources across three different build systems...

malfet replied:
You have to modify build system anyway, to add dependent libraries and so on. Also, making a typo in those build definitions leaves the test disabled forever, which is very hard to find compared to compile/link errors.
If you insist of #ifdef guards, wouldn't it be better to make them negative (i.e. compile the code unless `DISABLE_TENSORPIPE` define is set)

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]

lw committed:
Update on "Guard TensorPipe agent by USE_TENSORPIPE"


The USE_TENSORPIPE flag was guarding some functions in utils.h/cpp but it wasn't guarding the agent itself, leading to compile errors. This should fix it?

Differential Revision: [D22978717](https://our.internmc.facebook.com/intern/diff/D22978717/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22978717/)!

[ghstack-poisoned]
